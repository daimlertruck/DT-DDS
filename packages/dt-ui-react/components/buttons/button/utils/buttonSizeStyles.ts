import { CustomTheme as Theme } from '../../../../types';
import { ButtonSizes } from '../constants/types';

export const buttonSizeStyles = (size: ButtonSizes, theme: Theme) =>
  ({
    large: `
      ${theme.fontStyles.button1}
      padding: ${theme.spacing['2xs']} ${theme.spacing.s};
    `,
    medium: `
      ${theme.fontStyles.button2}
      padding: ${theme.spacing['3xs']} ${theme.spacing.xs};
    `,
    small: `
      ${theme.fontStyles.button2}
      padding: ${theme.spacing['4xs']} ${theme.spacing['2xs']};
    `,
  }[size]);
