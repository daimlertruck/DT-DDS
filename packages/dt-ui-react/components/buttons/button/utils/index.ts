export * from './buttonVariantColors';
export * from './buttonVariantStyles';
