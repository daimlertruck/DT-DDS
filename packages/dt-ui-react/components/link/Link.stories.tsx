import { Story } from '@storybook/react';

import { LinkTextSize } from './constants';

import { Link, LinkProps } from '.';

export default {
  title: 'Data Display/Link',
  component: Link,
};

const Template: Story<LinkProps> = ({ ...props }) => {
  return <Link {...props} />;
};

export const Default = Template.bind({});

Default.args = {
  children: 'Redirect to',
  href: 'https://www.google.com/',
  target: '_blank',
  textSize: LinkTextSize.Medium,
  isUnderline: true,
  isDisabled: false,
};
