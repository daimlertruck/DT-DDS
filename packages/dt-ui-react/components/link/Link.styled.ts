import styled from '@emotion/styled';

import { ButtonStyled } from '../buttons/button/Button.styled';

import { LinkTextSize } from './constants';

interface LinkStyledProps {
  isUnderline: boolean;
  disabled?: boolean;
  textSize?: string;
}

export const LinkStyled = styled.a<LinkStyledProps>`
  ${({ theme, disabled, isUnderline, textSize }) => `
    color: ${theme.palette.alpha.primary};
    text-decoration: ${isUnderline ? 'underline' : 'none'};

    ${
      disabled &&
      `
        opacity: 0.4;
        pointer-events: none;
    
      `
    };

    ${
      textSize == LinkTextSize.Large &&
      `
        ${theme.fontStyles.alpha.p}
      `
    };

    ${
      textSize == LinkTextSize.Medium &&
      `
        ${theme.fontStyles.alpha.pSmall}
      `
    };

    ${
      textSize == LinkTextSize.Small &&
      `
        ${theme.fontStyles.alpha.pXXSmall}
      `
    };
  `}
`;

export const ButtonLinkStyled = ButtonStyled.withComponent('a');
