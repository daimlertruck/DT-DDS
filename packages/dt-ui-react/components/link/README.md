# Link

A Link will redirect the user somewhere (different page or location in the same page), open the email provider with a default receiver, or download a file.

## Usage

```jsx
import { Link } from '@dt-ui/react';

export const App = () => {
  return <Link href='/'>Some Link</Link>;
};
```

## API

### Link

| Property      | Type               | Default               | Description                                                      |
| ------------- | ------------------ | --------------------- | ---------------------------------------------------------------- |
| `children`    | `ReactNode`        | -                     | Child components to be rendered                                  |
| `dataTestId`  | `string`           | tag                   | Customizable test identifier                                     |
| `onClick`     | `function`         | -                     | The triggered function when the Link is clicked                  |
| `isUnderline` | `boolean`          | true                  | Determines if the text is underlined or not                      |
| `isDisabled`  | `boolean`          | `false`               | Specifies if the element should be disabled                      |
| `textSize`    | `LinkTextSize`     | `LinkTextSize.Medium` | Specifies the text size of the element                           |
| `type`        | `button` \| `text` | `text`                | If set to true, the component inherits the Button css properties |

This component also inherits the native anchor properties. You can know more about this [here](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/a).
