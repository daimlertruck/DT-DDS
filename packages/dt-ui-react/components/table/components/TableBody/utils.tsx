import { Children, ReactNode, cloneElement, isValidElement } from 'react';

import { TableDataCellHeader } from '../TableDataCell/TableDataCell';

export const addHeadersToDataCells = (
  dataCells: ReactNode,
  headers: ReactNode[]
) => {
  return Children.map(dataCells, (dataCell, index) => {
    if (isValidElement(dataCell)) {
      const header = headers[index];
      return cloneElement(
        dataCell,
        {},
        <>
          <TableDataCellHeader>{header}</TableDataCellHeader>
          {dataCell.props.children}
        </>
      );
    }
    return dataCell;
  });
};

export const addHeadersToDataCellsInRows = (
  rows: ReactNode,
  headers: ReactNode[]
) => {
  return Children.map(rows, (row) => {
    if (isValidElement(row)) {
      return cloneElement(
        row,
        {},
        addHeadersToDataCells(row.props.children, headers)
      );
    }
    return row;
  });
};
