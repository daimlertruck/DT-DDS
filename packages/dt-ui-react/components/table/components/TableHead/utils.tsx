import { Children, ReactNode, isValidElement } from 'react';

const extractColumnHeaders = (columnHeaders: ReactNode): ReactNode[] => {
  return Children.toArray(columnHeaders).reduce<ReactNode[]>(
    (headers, columnHeader) => {
      if (isValidElement(columnHeader)) {
        return [...headers, columnHeader.props.children];
      }

      return headers;
    },
    []
  );
};

export const extractTableHeaders = (theadRow: ReactNode) => {
  if (isValidElement(theadRow)) {
    return extractColumnHeaders(theadRow.props.children);
  }

  return [];
};
