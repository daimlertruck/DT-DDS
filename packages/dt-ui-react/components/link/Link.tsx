import { forwardRef } from 'react';

import { BaseProps } from '../../types';
import { ButtonSize } from '../buttons/button/constants';

import { LinkTextSize } from './constants';
import { LinkType } from './constants/types';
import { ButtonLinkStyled, LinkStyled } from './Link.styled';

export interface LinkProps
  extends BaseProps,
    React.AnchorHTMLAttributes<HTMLAnchorElement> {
  isUnderline?: boolean;
  isDisabled?: boolean;
  onClick?: () => void;
  textSize?: string;
  type?: LinkType;
}

export const Link = forwardRef<HTMLAnchorElement, LinkProps>(
  (
    {
      isDisabled,
      children,
      isUnderline = true,
      textSize = LinkTextSize.Medium,
      dataTestId,
      onClick,
      style,
      type = 'text',
      ...rest
    },
    ref
  ) => {
    return (
      <>
        {type === 'button' ? (
          <ButtonLinkStyled
            data-testid={dataTestId ?? 'link-button'}
            variant='solid'
            size={textSize as ButtonSize}
            ref={ref}
            style={style}
            {...rest}
            color='primary'
          >
            {children}
          </ButtonLinkStyled>
        ) : (
          <LinkStyled
            textSize={textSize}
            ref={ref}
            data-testid={dataTestId ?? 'link'}
            onClick={onClick}
            disabled={isDisabled}
            isUnderline={isUnderline}
            style={style}
            {...rest}
          >
            {children}
          </LinkStyled>
        )}
      </>
    );
  }
);
