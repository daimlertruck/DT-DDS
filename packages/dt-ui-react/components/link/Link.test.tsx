import { render } from '@testing-library/react';
import React from 'react';

import { withProviders } from '../../utils';
import { Link } from '../link';

describe('<Link /> component', () => {
  const ProvidedLink = withProviders(Link);

  it('Should render a html link element with the correct title', () => {
    const { container } = render(
      <ProvidedLink href='/'>Redirect To</ProvidedLink>
    );

    expect(container).toMatchSnapshot();
  });

  it('Should render a html underlined link element with the correct title', () => {
    const { container } = render(
      <ProvidedLink href='/' isUnderline={true}>
        Redirect To
      </ProvidedLink>
    );

    expect(container).toMatchSnapshot();
  });

  it('Should render a disabled html link element', () => {
    const { container } = render(
      <ProvidedLink href='/' isDisabled={true}>
        Redirect To
      </ProvidedLink>
    );

    expect(container).toMatchSnapshot();
  });

  it('Should render an anchor element with button styles', () => {
    const { container } = render(
      <ProvidedLink href='/' type='button'>
        Redirect To
      </ProvidedLink>
    );

    expect(container).toMatchSnapshot();
  });
});
