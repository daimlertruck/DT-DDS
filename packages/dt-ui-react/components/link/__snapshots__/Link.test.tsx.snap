// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Link /> component Should render a disabled html link element 1`] = `
.emotion-0 {
  color: #000000;
  -webkit-text-decoration: underline;
  text-decoration: underline;
  opacity: 0.4;
  pointer-events: none;
  font-weight: 400;
  font-size: 16px;
  line-height: 21.44px;
  letter-spacing: normal;
  margin: 0;
  text-transform: none;
}

<div>
  <a
    class="emotion-0 emotion-1"
    data-testid="link"
    disabled=""
    href="/"
  >
    Redirect To
  </a>
</div>
`;

exports[`<Link /> component Should render a html link element with the correct title 1`] = `
.emotion-0 {
  color: #000000;
  -webkit-text-decoration: underline;
  text-decoration: underline;
  font-weight: 400;
  font-size: 16px;
  line-height: 21.44px;
  letter-spacing: normal;
  margin: 0;
  text-transform: none;
}

<div>
  <a
    class="emotion-0 emotion-1"
    data-testid="link"
    href="/"
  >
    Redirect To
  </a>
</div>
`;

exports[`<Link /> component Should render a html underlined link element with the correct title 1`] = `
.emotion-0 {
  color: #000000;
  -webkit-text-decoration: underline;
  text-decoration: underline;
  font-weight: 400;
  font-size: 16px;
  line-height: 21.44px;
  letter-spacing: normal;
  margin: 0;
  text-transform: none;
}

<div>
  <a
    class="emotion-0 emotion-1"
    data-testid="link"
    href="/"
  >
    Redirect To
  </a>
</div>
`;

exports[`<Link /> component Should render an anchor element with button styles 1`] = `
.emotion-0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
  cursor: pointer;
  width: 100%;
  min-width: 80px;
  -webkit-transition: background-color 0.2s ease-in-out,color 0.2s ease-in-out,border-color 0.2s ease-in-out;
  transition: background-color 0.2s ease-in-out,color 0.2s ease-in-out,border-color 0.2s ease-in-out;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  border: 2px solid transparent;
  gap: 8px;
  border-radius: 0px;
  font-weight: 700;
  font-size: 0.875rem;
  line-height: 1.25rem;
  letter-spacing: normal;
  margin: 0;
  text-transform: none;
  padding: 12px 24px;
  border-color: #000000;
  color: #FFFFFF;
  background-color: #000000;
}

.emotion-0:hover {
  color: #FFFFFF;
  background-color: #454545;
  border-color: #454545;
}

.emotion-0:disabled {
  cursor: not-allowed;
  color: #FFFFFF;
  background-color: #DEDEDE;
  border-color: #DEDEDE;
}

.emotion-0:focus-visible {
  outline: 3px solid #131313;
  outline-offset: 1px;
}

@media only screen and (min-width: 768px) {
  .emotion-0 {
    width: auto;
  }
}

<div>
  <a
    class="emotion-0 emotion-1"
    color="primary"
    data-testid="link-button"
    href="/"
  >
    Redirect To
  </a>
</div>
`;
